### PR TITLE
fix: add migration for bad fragment bitmaps

### PR DESCRIPTION
**Pull request by westonpace**

### CODE REVIEW

File: rust/lance-core/src/format/manifest.rs
Comment (chebbyChefNEQ) [inline rust/lance-core/src/format/manifest.rs:204]: any reason we don't want to use the feature flag system here?

Comment (westonpace) [reply rust/lance-core/src/format/manifest.rs:204]: What's that?

Comment (westonpace) [reply rust/lance-core/src/format/manifest.rs:204]: Or, more accurately, how does that apply?

Comment (westonpace) [reply rust/lance-core/src/format/manifest.rs:204]: Ah, I think I see what you mean.  I was thinking "rust features" but I think you mean the lance writer features.

Instead of relying on the writer version we could look for `writer_feature_flags`.  I'm not sure.  Technically, there isn't a new proto feature here.  We didn't make any changes to how the fragment bitmap is stored in the proto.  The problem is specific to our implementation of Lance.  E.g. if some other writer was creating lance files then they should be fine.

It also feels a little weird to have a writer feature named "no really, you can trust the fragment bitmap now"

CC @wjones127 for second opinion

Comment (wjones127) [reply rust/lance-core/src/format/manifest.rs:204]: What Weston saying is exactly what I was thinking. It's possible in the future we will have multiple reader / writer implementations, and I want those flags to be meaningful to them.

File: rust/lance/src/io/commit.rs
Comment (wjones127) [inline rust/lance/src/io/commit.rs:263]: Perhaps we should have a fast path, where we see if [index.dataset_version](https://github.com/lancedb/lance/blob/48b19829d3c7bd07ebadb3ebc1c3b7eb2b6c9068/protos/index.proto#L37C16-L37C16) still exists and if it does load the fragments from that manifest? Does that seem like it would be faster than loading from the index?

Comment (westonpace) [reply rust/lance/src/io/commit.rs:277]: Loading the fragments from that manifest could be incorrect because we update the dataset version when we optimize indices.  So we could have:

TRAIN ON DS VERSION 1 FRAGS 0,1,2,3
DELETE FRAG 0
OPTIMIZE TO DS VERSION 2 FRAGS 1, 2, 3

Now, `index.dataset_version` is `2` which points to frags `1, 2, 3` but the index actually covers frags `0, 1, 2, 3`.

File: rust/lance/src/dataset.rs
Comment (westonpace) [reply rust/lance/src/dataset.rs:1]: Added

File: rust/lance/src/io/commit.rs
Comment (westonpace) [inline rust/lance/src/io/commit.rs:266]: This is unfortunate because we now have to pay this cost on every commit.  Fortunately, it is probably fairly rare that commits involve indices (and those operations are probably already kind of slow).

Perhaps someday in the future, when we feel confident no one is using these old versions, we can pull this check back out.

File: rust/lance/src/io/commit.rs
Comment (wjones127) [reply rust/lance/src/io/commit.rs:277]: That makes sense. Out of curiosity: Do we update the UUID of the index when we optimize the index? That would let us detect whether the original fragments can be used.

I'm not sure we could use that if we could; it's just extra IOs for the tables that need fixing.

File: rust/lance/src/io/commit.rs
Comment (wjones127) [inline rust/lance/src/io/commit.rs:264]: I'm worried about this code path. Datasets that regularly run `cleanup_old_versions()` and `optimize_indices()` will hit this a lot, won't they? Maybe we should actually attach a writer version to the indices themselves? That way we can tell whether we can trust the index even if the version it was originally created in is long gone. Indexes might be quite long-lived, right?

File: rust/lance/src/io/commit.rs
Comment (westonpace) [reply rust/lance/src/io/commit.rs:277]: The UUID changes anytime the index changes (compact or update/optimize) though I'm not sure how that helps?

Comment (wjones127) [reply rust/lance/src/io/commit.rs:277]: It provides an algorithm to find if an index has never been optimized:

1. Checkout `dataset_version`
2. Compare the `UUID` of index versus the `UUID` in `dataset_version`. If they are the same, the index has never been optimized. Otherwise, it has been optimized / compacted.

Comment (westonpace) [reply rust/lance/src/io/commit.rs:277]: Ok, that's pretty clever.  Though I'll probably hold off as I don't expect this recalculate to run that often and it isn't *that* expensive.

File: rust/lance/src/io/commit.rs
Comment (westonpace) [reply rust/lance/src/io/commit.rs:264]: > Datasets that regularly run cleanup_old_versions() and optimize_indices() will hit this a lot, won't they?

Yes

> Maybe we should actually attach a writer version to the indices themselves?

Let me give that a try real quick

> Indexes might be quite long-lived, right?

Yes (not any single UUID but the conceptual index yes)

File: rust/lance/src/io/commit.rs
Comment (wjones127) [reply rust/lance/src/io/commit.rs:277]: Yeah, I agree it's not worth implementing for this issue. But something to keep in mind we *can* do, if it ever comes up again.

**Comment by westonpace** [comment]
@wjones127 so, as I was adding the index metadata's writer_version, I realized that we actually rewrite the index metadata on every commit.  This means we should detect and fix the broken fragment bitmaps on the very first commit of 0.8.15.  This means we can just use the dataset's `writer_version` and it will keep the fast path fast.

**Review by wjones127 — APPROVED** [review]
Great work!

**Pull request merged**
